java-checks/src/main/java/org/sonar/java/checks/MismatchPackageDirectoryCheck.java

```diff
@@ -41,20 +42,58 @@ public void scanFile(JavaFileScannerContext context) {
   @Override
   public void visitCompilationUnit(CompilationUnitTree tree) {
     PackageDeclarationTree packageDeclaration = tree.packageDeclaration();
-    if (packageDeclaration != null) {
-      String packageName = PackageUtils.packageName(packageDeclaration, File.separator);
-      File javaFile = context.getInputFile().file();
-      String dir = javaFile.getParent();
-      if (!dir.endsWith(packageName)) {
-        String dirWithoutDots = dir.replace(".", File.separator);
-        String issueMessage = MessageFormat.format(MESSAGE, javaFile.getName(), packageName, dir);
-        if (dirWithoutDots.endsWith(packageName)) {
-          context.reportIssue(this, packageDeclaration.packageName(), issueMessage + "(Do not use dots in directory names).");
-        } else {
-          context.reportIssue(this, packageDeclaration.packageName(), issueMessage + ".");
-        }
-      }
+    if (packageDeclaration == null) {
+      return;
     }
+
+    String packageName = PackageUtils.packageName(packageDeclaration, File.separator);
+    File fileDirectory = context.getInputFile().file().getParentFile();
+    String fileDirectoryPath = fileDirectory.getPath();
+    boolean packageNameIsSuffixOfDirPath = fileDirectoryPath.endsWith(packageName);
+
+    // In this case, path match package name.
+    if (packageNameIsSuffixOfDirPath) {
+      return;
+    }
+
+    File rootProjectWorkingDirectory = null;
+    try {
+      rootProjectWorkingDirectory = context.getRootProjectWorkingDirectory();
+    } catch (NullPointerException ignored) {
+      // RootProjectWorkingDirectory is initialized to null.
+      // NullPointerExceptions should not be thrown when accessing rootDirectory, but for now some context do it. See: SONARJAVA-5158.
```

Comment:
I am not sure where in the ticket an explanation is provided for the contexts where an issue is raised. Can we be more explicit and self-contained here?